File: desktop-client/src/app/game/services/game-state/game-state.service.ts
import { Injectable } from '@angular/core';
import { QuestionType } from '@common/enum/question-type';
import { PlayerInfo } from '@common/interfaces/player-info';
import { Question } from '@common/interfaces/question';

@Injectable({
    providedIn: 'root',
})
export class GameStateService {
    isSubmitted: boolean = false;
    playerScore: number = 0;
    playerInfo: PlayerInfo[];
    currentQuestion: Question;
    questionsLength: number = 0;
    currentQuestionIndex: number = 0;
    isEliminated = false;
    playersAwsKeys: string[] = [];

    get questionChoices(): string[] {
        return this.currentQuestion.choices.map((choice) => choice.text);
    }

    get questionIndexes(): number[] {
        return this.currentQuestion.choices.map((_, i) => i);
    }

    get questionType(): QuestionType {
        return this.currentQuestion.type as QuestionType;
    }

    resetGameState() {
        this.isSubmitted = false;
        this.isEliminated = false;
        this.playerScore = 0;
        this.currentQuestionIndex = 0;
        this.questionsLength = 0;
        this.currentQuestion = {} as Question;
    }

    changeQuestion(question: Question) {
        this.isSubmitted = false;
        this.currentQuestion = question;
        this.currentQuestionIndex++;
    }
}

File: mobile_client/lib/services/game/game_state_service.dart
import 'package:flutter/material.dart';
import 'package:mobile_client/constants/enum/question_type.dart';
import 'package:mobile_client/models/multi_choice.dart';
import 'package:mobile_client/models/player_info.dart';
import 'package:mobile_client/models/question.dart';

class GameStateService {
  ValueNotifier<bool> isSubmitted = ValueNotifier<bool>(false);
  ValueNotifier<int> playerScore = ValueNotifier<int>(0);
  ValueNotifier<Question?> currentQuestion = ValueNotifier<Question?>(null);
  int currentQuestionIndex = 0;
  int questionsLength = 0;
  List<PlayerInfo> playerInfos = [];
  List<String> playersAwsKeys = [];
  ValueNotifier<bool> isEliminated = ValueNotifier<bool>(false);

  List<MultiChoice> get questionChoices =>
      currentQuestion.value?.choices.map((choice) => choice).toList() ?? [];

  List<int> get questionIndexes => currentQuestion.value?.choices != null
      ? List.generate(currentQuestion.value!.choices.length, (i) => i)
      : [];

  QuestionType get questionType => currentQuestion.value!.type;

  void submit() {
    isSubmitted.value = true;
  }

  void resetGameState() {
    isEliminated.value = false;
    isSubmitted.value = false;
    playerScore.value = 0;
    currentQuestionIndex = 0;
    questionsLength = 0;
  }

  void changeQuestion(Question question) {
    isSubmitted.value = false;
    currentQuestion.value = question;
    currentQuestionIndex++;
  }
}
